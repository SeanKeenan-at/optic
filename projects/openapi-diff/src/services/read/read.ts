import { ISpecReader, OpenAPIDiffingQuestions } from './types';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
import { openApiQueries } from './queries';
import { IFilePatch } from '../patch/types';
import fs from 'fs-extra';
import { defaultEmptySpec } from './debug-implementations';
import {
  ParseOpenAPIResult,
  parseOpenAPIWithSourcemap,
} from '@useoptic/openapi-io';

export function createOpenApiFileSystemReader(filePath: string): ISpecReader {
  const start = Date.now();
  let end = Date.now();
  let result: Promise<ParseOpenAPIResult> | undefined;

  let questionsPromise: Promise<OpenAPIDiffingQuestions> | undefined;

  function reload() {
    result = new Promise(async (resolve) => {
      const checkIfExists = await fs.pathExists(filePath);
      if (checkIfExists) {
        return resolve(await parseOpenAPIWithSourcemap(filePath));
      } else {
        await fs.writeJSON(filePath, defaultEmptySpec);
        return resolve(await parseOpenAPIWithSourcemap(filePath));
      }
    });

    result.finally(() => {
      end = Date.now();
    });

    questionsPromise = result.then((parsed) => openApiQueries(parsed.jsonLike));
  }

  reload();

  return {
    rootFile(): string {
      return filePath;
    },
    describeLocation(): string {
      return `${filePath.replace(process.cwd(), '')}`;
    },
    save: async (patch: IFilePatch): Promise<void> => {
      await Promise.all(
        patch.files.map(async (i) => fs.writeFile(i.path, i.newContents))
      );
      reload();
    },
    flattenedSpecification: async (): Promise<OpenAPIV3.Document> => {
      const { jsonLike } = await result;
      return jsonLike;
    },
    didLoad: async () => {
      // wait for it to finish
      await new Promise((resolve) => {
        result.finally(() => resolve(undefined));
      });

      try {
        await result;
        return { success: true, durationMillis: end - start };
      } catch (e: any) {
        return {
          success: false,
          error: e.message,
          durationMillis: end - start,
        };
      }
    },
    isStale: async () => {
      return false; // @todo checksum dependencies
    },
    questions: async (): Promise<OpenAPIDiffingQuestions> => {
      return questionsPromise;
    },
  };
}
