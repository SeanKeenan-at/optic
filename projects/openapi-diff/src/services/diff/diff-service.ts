import { IDiff, IDiffService } from './types';
import { OpenAPIDiffingQuestions } from '../read/types';
import { ApiTraffic } from '../traffic/types';
import { shouldDiffAgainstThisSpec } from './differs/should-diff';
import { urlPathDiff, urlPathDiffFromSpec } from './differs/url-path-diff';
import { JsonSchemaJsonDiffer } from './differs/json-schema-json-diff/types';
import { opticJsonSchemaDiffer } from './differs/json-schema-json-diff';
import { pathMethodOperationDiffer } from './differs/path-method-operation-diff';
import { responsesDiffer } from './differs/responses';

export function createDiffServiceWithCachingProjections(
  spec: OpenAPIDiffingQuestions,
  jsonSchemaDiffer: JsonSchemaJsonDiffer = opticJsonSchemaDiffer()
): IDiffService {
  const shouldDiff = shouldDiffAgainstThisSpec(spec);
  const pathMatcher = urlPathDiffFromSpec(spec);
  const operationMatcher = pathMethodOperationDiffer(spec);
  const responsesMatcher = responsesDiffer(spec, jsonSchemaDiffer);

  return {
    jsonSchemaDiffer,
    compare: async (
      traffic: ApiTraffic
    ): Promise<{ diffs: IDiff[]; errors: string[] }> => {
      /*
        This is nested, feels ugly, but probably is the best way to represent a nested data structure that can be diffed...for now
        open to suggestions
       */
      const continueWithDiff =
        shouldDiff.responseDiffsForTraffic(traffic).isMatch;
      if (!continueWithDiff) return { diffs: [], errors: [] };

      const matchesPath = pathMatcher.compareToPath(
        traffic.method,
        traffic.path
      );

      // here we know we've matched an operation
      if (matchesPath.isMatch) {
        const matchesOperation =
          operationMatcher.comparePathAndMethodToOperation(
            matchesPath.context,
            traffic.method
          );

        if (matchesOperation.isMatch) {
          const results: { diffs: IDiff[]; errors: string[] } = {
            diffs: [],
            errors: [],
          };
          // match the response
          const response = responsesMatcher.responseDiffsForTraffic(
            traffic,
            matchesOperation.context
          );

          if (response.isMatch) {
            // response content & schema diff
            const responseContentMatch =
              responsesMatcher.responseContentDiffsForTraffic(
                traffic,
                response.context
              );

            results.diffs.push(...responseContentMatch.diffs);
          } else results.diffs.push(...response.diffs);

          // collect diffs from failed matches
          return results;
        } else return { diffs: matchesOperation.diffs, errors: [] };
      }
      return { diffs: matchesPath.diffs, errors: [] };
    },
  };
}
