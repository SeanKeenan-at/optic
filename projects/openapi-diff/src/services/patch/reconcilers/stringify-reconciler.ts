import { IFilePatch, IPatchOpenAPIReconciler } from '../types';
import { ISpecReader } from '../../read/types';
import {
  IPatchGroup,
  PatchesToSave,
} from '../incremental-json-patch/json-patcher';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
import stringify from 'json-stable-stringify';

export function StringifyReconciler(
  reader: ISpecReader
): IPatchOpenAPIReconciler {
  return {
    patchesToFileMutations: async (
      patchesToSave: PatchesToSave<OpenAPIV3.Document>
    ): Promise<IFilePatch> => {
      // console.log(JSON.stringify(patchesToSave.document, null, 2));
      return {
        files: [
          {
            path: reader.rootFile(),
            newContents: stringify(patchesToSave.document, { space: 1 }),
          },
        ],
      };
    },
  };
}
