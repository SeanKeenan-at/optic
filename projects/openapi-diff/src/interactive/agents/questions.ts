/*
  Agents are presented with questions, and must decide how to answer them. Possible answers
  - see a new url | add a new path, ignore the path/method pattern
  - see a shape diff | extend spec, replace, type
  - see a new body   | add it, ignore it
 */

import { ApiTraffic } from '../../services/traffic/types';
import { DiffType, IDiff, UnmatchedPath } from '../../services/diff/types';
import { v4 as uuidv4 } from 'uuid';

export type QuestionsForAgent = AddPathQuestion;

export type QuestionsForAgentAnswered = QuestionsForAgent & {
  answer: QuestionsForAgent['answer'];
};

interface Questions<A> {
  type: AnswerQuestionTypes;
  answer?: A;
  uuid: string;
}

export enum AnswerQuestionTypes {
  AddPath,
}

// Path add  Q / A
export interface AddPathQuestion extends Questions<AddPathAnswer> {
  type: AnswerQuestionTypes.AddPath;
  diff: UnmatchedPath;
  example: ApiTraffic;
}
export interface AddPathAnswer {
  pathPattern: string;
}
