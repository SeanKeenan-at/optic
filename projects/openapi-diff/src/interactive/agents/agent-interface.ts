import { AnswerQuestionTypes, QuestionsForAgent } from './questions';
import { ApiTraffic } from '../../services/traffic/types';
import { IDiff, IDiffService } from '../../services/diff/types';
import { IPatchOpenAPI } from '../../services/patch/types';

export interface AgentContext {}

export enum AgentEventEnum {
  DiffReceived = 'DiffReceived',
  Reset = 'Reset',
  AnswerQuestion = 'AnswerQuestion',
  SkipInteraction = 'SkipInteraction',
  SkipQuestion = 'SkipQuestion',
  AdvancePastError = 'AdvancePastError',
  Start = 'Start',
  Finish = 'Finish',
}

export type ProcessDiffContext = AgentContext & {
  diffs: IDiff[];
  traffic: ApiTraffic;
};
export type WaitingOnInputDiffContext = AgentContext & {
  diffs: IDiff[];
  questions: QuestionsForAgent[];
  traffic: ApiTraffic;
};

export type AgentEvent =
  | { type: AgentEventEnum.DiffReceived; diffs: IDiff[]; traffic: ApiTraffic }
  | { type: AgentEventEnum.AnswerQuestion; id: string; answer: any }
  | { type: AgentEventEnum.Reset }
  | { type: AgentEventEnum.SkipInteraction }
  | { type: AgentEventEnum.SkipQuestion };

export type AgentTypestate =
  | {
      value: 'idle';
      context: AgentContext;
    }
  | {
      value: 'process_diffs';
      context: ProcessDiffContext;
    }
  | {
      value: 'waiting_for_input';
      context: WaitingOnInputDiffContext;
    }
  | {
      value: 'error';
      context: AgentContext & { error: string };
    }
  | {
      value: 'finished';
      context: AgentContext & {};
    };

export interface AgentIntent {
  name: string;
  handleDiffs: (
    diff: IDiff,
    example: ApiTraffic,
    patch: IPatchOpenAPI,
    diffService: IDiffService,
    askQuestion: (question: QuestionsForAgent) => void
  ) => void;
  applyAnswerAsPatch: (
    questionAnswer: QuestionsForAgent,
    patch: IPatchOpenAPI
  ) => void;
}
