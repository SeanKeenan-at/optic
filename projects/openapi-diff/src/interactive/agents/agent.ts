import { assign, createMachine, interpret } from 'xstate';
import {
  AgentContext,
  AgentEvent,
  AgentEventEnum,
  AgentIntent,
  AgentTypestate,
  ProcessDiffContext,
  WaitingOnInputDiffContext,
} from './agent-interface';
import isEqual from 'lodash.isequal';
import { InteractiveDiffMachineType } from '../machine';
import { IDiff } from '../../services/diff/types';
import { QuestionsForAgent } from './questions';
import { ApiTraffic } from '../../services/traffic/types';
import { DiffEventEnum } from '../machine-interface';

export function createAgentMachine(
  diffMachine: InteractiveDiffMachineType,
  intent: AgentIntent
) {
  const agentMachine = createMachine<AgentContext, AgentEvent, AgentTypestate>({
    id: 'agent',
    initial: 'idle',
    context: {},
    states: {
      idle: {
        entry: assign({}),
        on: {
          [AgentEventEnum.DiffReceived]: {
            target: 'process_diffs',
            actions: assign((context, event) => {
              return {
                diffs: event.diffs,
                traffic: event.traffic,
                questions: [],
              };
            }),
          },
        },
      },
      process_diffs: {
        invoke: {
          // this entire src could be made a service for other kinds of diff intents
          src: async (
            ctx,
            event
          ): Promise<{ questions: QuestionsForAgent[] }> => {
            const { diffs, traffic } = ctx as ProcessDiffContext;

            // for user input
            const questions: QuestionsForAgent[] = [];
            // use this to patch diffs this intent automates
            const { specInterface, diffService } = diffMachine.state.context;

            diffs.forEach((diff) =>
              intent.handleDiffs(
                diff,
                traffic,
                specInterface.patch,
                diffService,
                (question) => {
                  questions.push(question);
                }
              )
            );

            // if automated patches are dispatched, we run these first
            // if (specInterface.patch.listPatches().length) {
            //   // diffMachine.send({
            //   //   type: DiffEventEnum.Agent_Submitted_Patch,
            //   // });
            //   return { questions: [] };
            // }

            return { questions };
          },
          onDone: [
            {
              target: 'check_should_flush',
              actions: assign((ctx, event) => {
                return {
                  questions: event.data.questions,
                };
              }),
            },
          ],
          onError: {
            actions: (ctx, event) => {
              console.log(event);
            },
          },
        },
      },
      waiting_for_input: {
        on: {
          [AgentEventEnum.SkipInteraction]: {
            actions: (ctx, event) => {
              diffMachine.send({
                type: DiffEventEnum.Agent_Skipped_Interaction,
              });
            },
            target: 'idle',
          },
          [AgentEventEnum.AnswerQuestion]: {
            actions: assign((ctx, event) => {
              const context = ctx as WaitingOnInputDiffContext;
              const { specInterface } = diffMachine.state.context;
              const question = context.questions.find(
                (i) => i.uuid === event.id
              );
              if (question) {
                question.answer = event.answer;
              }
              return { questions: context.questions };
            }),
            target: 'check_should_flush',
          },
        },
      },
      check_should_flush: {
        invoke: {
          src: async (ctx) => {
            const context = ctx as WaitingOnInputDiffContext;
            const { specInterface } = diffMachine.state.context;

            const result = { shouldIdle: true, shouldFlush: false };

            if (
              context.questions.length > 0 &&
              context.questions.every((i) => i.answer)
            ) {
              context.questions.forEach((question) =>
                intent.applyAnswerAsPatch(question, specInterface.patch)
              );
              result.shouldFlush = true;
            }

            if (
              context.questions.length > 0 &&
              context.questions.some((i) => !i.answer)
            ) {
              result.shouldFlush = false;
              result.shouldIdle = false;
            }

            if (
              context.questions.length === 0 &&
              specInterface.patch.listPatches().length > 0
            ) {
              result.shouldFlush = true;
            }

            return result;
          },
          onDone: [
            {
              target: 'idle',
              actions: () => {
                diffMachine.send({ type: DiffEventEnum.Agent_Submitted_Patch });
              },
              cond: (ctx, event) => event.data.shouldFlush,
            },
            {
              target: 'idle',
              cond: (ctx, event) => event.data.shouldIdle,
            },
            { target: 'waiting_for_input' },
          ],
          onError: {
            actions: [
              (ctx, error) => {
                console.error(error);
                diffMachine.send({
                  type: DiffEventEnum.Agent_Skipped_Interaction,
                });
              },
            ],
            target: 'idle',
          },
        },
      },
    },
  });

  const service = interpret(agentMachine);

  service.start();

  // response to state changes of diff machine
  let previousAsk: { diffs: IDiff[]; traffic: ApiTraffic } | undefined;

  diffMachine.onTransition((state, event) => {
    // transitioned to waiting_for_input

    if (state.value === 'waiting_for_input') {
      if (
        !isEqual(previousAsk, {
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        })
      ) {
        service.send({
          type: AgentEventEnum.DiffReceived,
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        });
        previousAsk = {
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        };
      }
    } else {
      previousAsk = undefined;
    }
  });

  return service;
}
