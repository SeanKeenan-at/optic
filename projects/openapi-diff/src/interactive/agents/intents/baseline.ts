import { AgentIntent } from '../agent-interface';
import { AnswerQuestionTypes } from '../questions';
import {
  DiffType,
  IDiffService,
  ShapeDiffTypes,
  UnmatchedPath,
  UnmatchedResponse,
} from '../../../services/diff/types';
import { v4 as uuidv4 } from 'uuid';
import { JsonSchemaPatch } from '../../../services/diff/differs/json-schema-json-diff/plugins/plugin-types';

export function baselineIntent(): AgentIntent {
  return {
    name: 'baseline',
    handleDiffs: (
      diff,
      example,
      patch,
      diffService: IDiffService,
      askQuestion
    ) => {
      switch (diff.type) {
        case DiffType.UnmatchedPath: {
          askQuestion({
            type: AnswerQuestionTypes.AddPath,
            uuid: uuidv4(),
            diff: diff as UnmatchedPath,
            example,
          });
          break;
        }
        case DiffType.UnmatchedMethod: {
          const { path, method } = diff as UnmatchedPath;
          patch.init.operation(path, method, example);

          break;
        }
        // handle property diffs
        case DiffType.BodyAdditionalProperty:
        case DiffType.BodyUnmatchedType:
        case DiffType.BodyMissingRequiredProperty: {
          const possiblePatches = diffService.jsonSchemaDiffer.diffToPatch(
            diff as ShapeDiffTypes,
            patch.forkedPatcher()
          );
          const extendingPatch = possiblePatches.find((i) => i.extends) as
            | JsonSchemaPatch
            | undefined;
          if (extendingPatch) {
            patch.patch.property(extendingPatch);
          }

          break;
        }
        case DiffType.UnmatchedResponse: {
          const { path, method, statusCode } = diff as UnmatchedResponse;
          patch.init.response(path, method, example);
          break;
        }
        default:
          console.warn(`diff type not handed by agent ${diff.type}`);
      }
    },
    applyAnswerAsPatch: (questionAnswer, patch) => {
      if (!questionAnswer.answer) throw new Error('no answer provided');

      switch (questionAnswer.type) {
        case AnswerQuestionTypes.AddPath:
          const addPath = questionAnswer;
          patch.init.operation(
            addPath.answer.pathPattern,
            addPath.diff.method,
            addPath.example
          );
          break;
      }
    },
  };
}
