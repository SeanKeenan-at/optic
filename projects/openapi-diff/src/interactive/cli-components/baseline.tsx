import React, { useRef } from 'react';
import { TrafficSource } from '../../services/traffic/types';
import { Box, Newline, Static, Text } from 'ink';
import { useIntentDrivenDiffMachine } from './hooks/use-intent-driven-diff-machine';
import { AnswerQuestionTypes, QuestionsForAgent } from '../agents/questions';
import { UnmatchedPathQuestion } from './questions/unmatched-path-question';
import { baselineIntent } from '../agents/intents/baseline';
import { WaitingOnInputDiffContext } from '../agents/agent-interface';
import { Log } from './ui/log';

export function Baseline(props: {
  source: TrafficSource;
  openApiFilePath?: string;
}) {
  const { state, context, send, log } = useIntentDrivenDiffMachine(
    props.source,
    baselineIntent,
    props.openApiFilePath
  );
  let dynamic = <></>;

  if (state === 'waiting_for_input') {
    const waitingForInputContext = context as WaitingOnInputDiffContext;
    const firstUnansweredQuestion = waitingForInputContext.questions.find(
      (i) => !i.answer
    );
    if (firstUnansweredQuestion) {
      dynamic = (
        <>
          {componentForQuestion(
            firstUnansweredQuestion,
            send.answer,
            send.skipInteraction
          )}
        </>
      );
    }
  }

  return (
    <Box flexDirection="column">
      <Log log={log} />
      {state === 'waiting_for_input' && (
        <Box flexDirection="column">{dynamic}</Box>
      )}
    </Box>
  );
}

function componentForQuestion(
  question: QuestionsForAgent | undefined,
  answerQuestion: (questionId: string, answer: any) => void,
  skip: () => void
) {
  if (!question) return null;

  switch (question.type) {
    case AnswerQuestionTypes.AddPath:
      return (
        <UnmatchedPathQuestion
          key={question.uuid}
          question={question}
          answerQuestion={answerQuestion}
          skipInteraction={skip}
        />
      );
    default:
      throw new Error(
        'no registered question / answer component for ' + question
      );
  }
}
