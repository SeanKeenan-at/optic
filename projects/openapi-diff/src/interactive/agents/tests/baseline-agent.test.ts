import { DebugSource } from '../../../services/traffic/sources/debug-implementations';
import { makeExample } from '../../../services/traffic/traffic/debug-simple';
import { newDiffMachine } from '../../machine';
import { OpenApiInterface } from '../../../services/openapi-read-patch-interface';
import { PassThroughSpecReader } from '../../../services/read/debug-implementations';
import { StringifyReconciler } from '../../../services/patch/reconcilers/stringify-reconciler';
import { waitFor } from '../../../utils/debug_waitFor';
import { createDiffServiceWithCachingProjections } from '../../../services/diff/diff-service';
import { ApiTraffic } from '../../../services/traffic/types';
import { createAgentMachine } from '../agent';
import { AgentEventEnum, WaitingOnInputDiffContext } from '../agent-interface';
import { baselineIntent } from '../intents/baseline';
import { AnswerQuestionTypes } from '../questions';

describe('baseline agents', () => {
  it('learn API one path at a time', async () => {
    const diffServiceHandlingTraffic = await exampleFixture([
      // ask me about this
      makeExample('/examples', 'get', '200', { items: ['world'] }),
      // do not ask me about this, just learn it
      makeExample('/examples', 'post', '200', { items: ['world'] }),
      // ask about this
      makeExample('/examples/example123', 'get', '200', {
        name: 'example',
        size: 'medium',
      }),
    ]);

    const machine = createAgentMachine(
      diffServiceHandlingTraffic,
      baselineIntent()
    );

    const pathPattenAnswers = ['/examples', '/examples/{exampleId}'];

    machine.onTransition((state) => {
      const context = state.context as WaitingOnInputDiffContext;
      if (state.value === 'waiting_for_input') {
        if (
          context.questions[0] &&
          context.questions[0].type === AnswerQuestionTypes.AddPath
        ) {
          const answer = pathPattenAnswers.shift();
          if (answer) {
            machine.send({
              type: AgentEventEnum.AnswerQuestion,
              id: context.questions[0].uuid,
              answer: {
                pathPattern: answer,
              },
            });
          }
        }
      }
    });
    await waitFor(500);
    const resultingSpec =
      await diffServiceHandlingTraffic.state.context.specInterface.read.flattenedSpecification();
    expect(resultingSpec).toMatchSnapshot();
  });

  it('automatically approve property diffs', async () => {
    const diffServiceHandlingTraffic = await exampleFixture([
      // baseline
      makeExample('/examples', 'get', '200', { field1: 'hello' }),
      // new field, and one of
      makeExample('/examples', 'get', '200', {
        field1: 123,
        field2: [{ size: 'sm' }, { size: 'xl' }],
      }),
      // new field made optional
      makeExample('/examples', 'get', '200', {
        field1: 123,
      }),
    ]);

    const machine = createAgentMachine(
      diffServiceHandlingTraffic,
      baselineIntent()
    );

    const pathPattenAnswers = ['/examples'];

    machine.onTransition((state) => {
      const context = state.context as WaitingOnInputDiffContext;
      if (state.value === 'waiting_for_input') {
        if (
          context.questions[0] &&
          context.questions[0].type === AnswerQuestionTypes.AddPath
        ) {
          const answer = pathPattenAnswers.shift();
          if (answer) {
            machine.send({
              type: AgentEventEnum.AnswerQuestion,
              id: context.questions[0].uuid,
              answer: {
                pathPattern: answer,
              },
            });
          }
        }
      }
    });

    await waitFor(500);
    const resultingSpec =
      await diffServiceHandlingTraffic.state.context.specInterface.read.flattenedSpecification();
    expect(resultingSpec).toMatchSnapshot();
  });

  it('automatically approve new response types', async () => {
    const diffServiceHandlingTraffic = await exampleFixture([
      // baseline
      makeExample('/examples', 'get', '200', { field1: 'hello' }),
      // new field, and one of
      makeExample('/examples', 'get', '404', {
        error: "it's broken",
      }),
    ]);

    const machine = createAgentMachine(
      diffServiceHandlingTraffic,
      baselineIntent()
    );

    const pathPattenAnswers = ['/examples'];

    machine.onTransition((state) => {
      const context = state.context as WaitingOnInputDiffContext;
      if (state.value === 'waiting_for_input') {
        if (
          context.questions[0] &&
          context.questions[0].type === AnswerQuestionTypes.AddPath
        ) {
          const answer = pathPattenAnswers.shift();
          if (answer) {
            machine.send({
              type: AgentEventEnum.AnswerQuestion,
              id: context.questions[0].uuid,
              answer: {
                pathPattern: answer,
              },
            });
          }
        }
      }
    });

    await waitFor(500);
    const resultingSpec =
      await diffServiceHandlingTraffic.state.context.specInterface.read.flattenedSpecification();
    expect(resultingSpec).toMatchSnapshot();
  });
});

async function exampleFixture(examples: ApiTraffic[]) {
  const source = new DebugSource(
    examples,
    // [
    // makeExample("/examples/123", "get", "200"),
    // makeExample("/examples/456", "get", "200"),
    // makeExample("/examples/789", "get", "200"),
    // makeExample("/examples/101112", "get", "200"),
    // makeExample("/examples", "post", "200"),
    // ],
    10
  );

  const diffService = newDiffMachine(
    source,
    () =>
      OpenApiInterface(new PassThroughSpecReader(), (reader) =>
        StringifyReconciler(reader)
      ),
    (spec) => createDiffServiceWithCachingProjections(spec),
    {
      maxQueue: 10,
    }
  );

  await source.start();

  return diffService;
}
