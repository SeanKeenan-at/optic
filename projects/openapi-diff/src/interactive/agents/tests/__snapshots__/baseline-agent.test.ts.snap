// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseline agents automatically approve new response types 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/examples": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "field1": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "field1",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
          "404": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "error": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "error",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline agents automatically approve property diffs 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/examples": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "field1": Object {
                      "oneOf": Array [
                        Object {
                          "type": "string",
                        },
                        Object {
                          "type": "number",
                        },
                      ],
                    },
                    "field2": Object {
                      "items": Object {
                        "properties": Object {
                          "size": Object {
                            "type": "string",
                          },
                        },
                        "required": Array [
                          "size",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "field1",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline agents learn API one path at a time 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/examples": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "items": Object {
                      "items": Object {
                        "type": "string",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "items",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
      "post": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "items": Object {
                      "items": Object {
                        "type": "string",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "items",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
    "/examples/{exampleId}": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "name": Object {
                      "type": "string",
                    },
                    "size": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "name",
                    "size",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [
        Object {
          "in": "path",
          "name": "exampleId",
          "required": true,
          "schema": Object {
            "type": "string",
          },
        },
      ],
    },
  },
}
`;
